tsweb/request_id.go

```diff
@@ -0,0 +1,62 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package tsweb
+
+import (
+	"context"
+	"net/http"
+)
+
+// RequestID is an opaque identifier for a HTTP request, used to correlate
+// user-visible errors with backend server logs. The RequestID is typically
+// threaded through an HTTP Middleware (WithRequestID) and then can be
+// extracted by HTTP Handlers to include in their logs.
+//
+// RequestID is an opaque identifier for a HTTP request, used to correlate
+// user-visible errors with backend server logs. If present in the context, the
+// RequestID will be printed alongside the message text and logged in the
+// AccessLogRecord.
+//
+// There is no particular format required for a RequestID, but ideally it should
+// be obvious to an end-user that it is something to record for support
+// purposes. One possible example for a RequestID format is:
+// REQ-{server identifier}-{timestamp}-{random hex string}.
+type RequestID string
+
+// RequestIDHeader is a custom HTTP header that the WithRequestID middleware
+// uses to determine whether to re-use a given request ID from the client
+// or generate a new one.
+const RequestIDHeader = "X-Tailscale-Request-Id"
+
+// WithRequestID is an http middleware that injects a RequestID in the
+// *http.Request Context. The value of that request id is either retrieved from
+// the RequestIDHeader or from the generator if the header does not exist. Inner
+// handlers can retrieve this ID from the RequestIDFromContext function.
+func WithRequestID(generator func(*http.Request) string) Middleware {
```

Comment:
Optional suggestion: Elsewhere, we follow the loose convention of using `WithFoo` to name functions that attach values to context, and while that's indirectly what this does, it's not quite that. Perhaps this could be `SetRequestID` instead so that if we want to export the function below, it could use the "conventional" name.

It's up to you though.

Reply:
Done 👍🏼